### ISSUE TITLE
MIPS32 and MIPS64 support

### ISSUE THREAD

**Issue by jasonwilliams200OK** [dotnet/coreclr#969]
Apple brought 64-bit support in ARM in 2011. Other vendors (Nvidia, Qualcomm) are bringing 64-bit ARM architecture in 2015. 

However, MIPS had 64-bit support since 1999. Linux and old Windows Mobile and Windows CE operating systems support MIPS for long time. But Windows 10 for *everything* doesn't support MIPS! We definitely live in the world of monopoly..

Would be really handy if CoreCLR and CoreFX folks along with microframework (https://github.com/NETMF/netmf-interpreter) continue to support for MIPS32 and 64 architectures for the love of technology widespread in open world.

Love for MIPS

Thank you.

**Comment by OtherCrashOverride** [comment]
I don't believe its an issue of ideology that MIPS is excluded, rather, its an issue of pragmatism: there are no customers targeting MIPS hardware.

CoreCLR is open source.  There is nothing preventing Imagination Technologies (the owners of MIPS) from adding support for their processors to it.  Imagination Tech is a large company with plenty of financial resources. There is no need for them to depend on "technological welfare" for support of their products.

**Comment by jasonwilliams200OK** [comment]
> there are no customers targeting MIPS hardware.

I suspect that you missed (not really) the news (anymore) that Android is supported on MIPS and there are MIPS based tablets running Android and the development is still ongoing of such devices.

**Comment by OtherCrashOverride** [comment]
Let me expand that a bit:
there are no [Microsoft] customers targeting MIPS hardware.

I am all for MIPS support. I am simply pointing out that current lack of support is not due to any evil agenda.  Further more, MIPS support is not dependent on Microsoft. Anyone, including Imagination Tech, can add it.  You should email or post to their forums your interest in using CoreCLR on devices containing their product.

**Comment by jasonwilliams200OK** [comment]
> there are no [Microsoft] customers targeting MIPS hardware.

How can you be so sure about that? I am a Microsoft customer and I own a MIPS-based tablet and would love to run .NET apps on my device.

> You should email or post to their forums your interest in using CoreCLR on devices containing their product.

You are mistaken. This is not how it works.

**Comment by jakesays** [comment]
@jasonwilliams200OK I'm guessing a Microsoft customer in this context means a product group within Microsoft. Also, can you explain how 'it' works?

**Comment by kangaroo** [comment]
@jasonwilliams200OK is 100% correct.  Thats not how it works.  How it works is if people are particularly passionate about a new platform or architecutre, they should submit pull requests.  I'm happy to review any and all PRs for MIPS support.  I have a MIPS box in the garage that I can validate on.

**Comment by benpye** [comment]
I'm guessing there isn't really any documentation on the process of doing JIT bringup on a new platform? MIPS would certainly be interesting, but I'm guessing not something MS would want to invest time in since it's not really targeted by any of their products. I'm not too sure what would have to be done to get the JIT working on a new architecture, from the current code base, though it would be impressive if done as a community project.

**Comment by terrajobst** [comment]
@kangaroo and @jasonwilliams200OK are correct. This is an open source project. It's by definition a collaboration with various individuals as well as companies. The code base you currently see is the actual product code that the CLR group at Microsoft produced over last 15 years. It's obviously based on the business interests that Microsoft had in the past.

As far as our future interest and thus investment goes: we already announced that we're going to support this runtime on non-Microsoft platforms. But pragmatically speaking, we (even though that may be hard to believe for some folks) don't have infinite resources to "simply" support CoreCLR everywhere. We do, however, want to work with others to port the code to more platforms. Our level of commitment will certainly vary based on roadmap, priorities, and overlap with other projects we're already working on.

That being said, it doesn't mean that we wouldn't accept PRs for platforms that Microsoft currently isn't interested in. It simply means that the community has to provide folks that are willing to maintain the architecture to a degree that it makes sense to be part of this repository. Not only is this pattern an option -- [it's highly desired](http://blogs.msdn.com/b/dotnet/archive/2015/07/14/first-net-port-award.aspx)!

We know for a fact that .NET is much bigger than just Microsoft. So please keep the suggestions for ports coming. There are many highly skilled and passionate people that care and are willing to collaborate -- as @kangaroo just proved by offering code reviews as well as validation.

**Comment by zezba9000** [comment]
I posted a suggestion in the .NET Foundation forums about Open Sourcing key parts of .NET Compact that has already has run on MIPS, PPC, ect for years: http://forums.dotnetfoundation.org/t/open-source-net-compact-runtime/1190

I think being able to use it as references source would greatly help the efforts of this topic.  If i'm right you might consider voicing this suggestion as well :)

**Comment by OtherCrashOverride** [comment]
> @kangaroo and @jasonwilliams200OK are correct.

> Not only is this pattern an option -- it's highly desired! [http://blogs.msdn.com/b/dotnet/archive/2015/07/14/first-net-port-award.aspx]

> There are many highly skilled and passionate people that care and are willing to collaborate -- as @kangaroo just proved by offering code reviews as well as validation.


Marcia, Marcia, Marcia! 

(Sorry, @kangaroo , couldn't resist. :wink: )
(For those not familiar with the reference: https://www.youtube.com/watch?v=w2fXs3bf-p0 "The Brady Bunch: Marcia, Marcia, Marcia! ")

**Comment by xied75** [comment]
For the record: https://en.wikipedia.org/wiki/Loongson
The 3A2000 model CPU is on sell on a mini-itx motherboard http://www.lemote.com/html/product/miniitx/2015/1230/28.html
The performance is on par of Intel Ivy-Bridge generation.

And it runs **Java**, **Go**, **Ruby**...

**Comment by jasonwilliams200OK** [comment]
@xied75, I started with Debian Jesse (MIPS) running on qemu-mips and converted AT&T `.S` assembly code in CoreCLR repo to its MIPS counterpart. There are couple of SIMD instructions which are different in (and some inherently not required by) MIPS which need some serious MIPS assembly darkarts which i lack.. If you have affluent knowledge about MIPS assembly, do contribute! :)

More on the devices; there are many interesting devices posted on their blog https://imgtec.com/blog/, most of them are SOC board, arduinos, and internet of things kinda solutions. And now there is even an All-in-one PC: http://www.pcworld.com/article/3040528/computers/this-russian-all-in-one-desktop-is-just-weird-enough-to-be-interesting.html --happened just around the 25th anniversary of 64-bit MIPS architecture and 35th year overall. :wink:

**Comment by benpye** [comment]
First step would be to get PAL working on MIPS, that should be fairly easy. Getting the JIT on MIPS would be quite a bit of work, I guess there is the interpreter fallback, which would be useful during bring up, but it would probably make sense to wait for RyuJIT to work well on x86 or ARM before trying to get another 32-bit platform working. I know there is MIPS64 but I'm pretty sure it'd a strict superset so a MIPS32 port seems to make more sense.

**Comment by jasonwilliams200OK** [comment]
@benpye, thanks! :)

For the first step:

* should we only configure cmake under PAL?
* can we run [PAL tests](https://github.com/dotnet/coreclr/blob/555cf28/Documentation/building/unix-test-instructions.md) as well without building JIT?
* any further advice on getting the JIT interpreter fallback up and running?

**Comment by benpye** [comment]
PAL and the PAL tests are pretty much standalone, they will work without the rest of coreclr. You will have to modify the top level CMakefile somewhat, I'd look through for the ARM defines and add suitable ones for MIPS. You could make it so that for MIPS only PAL was built.

To get the JIT up and running you'd have to ask @dotnet/jit-contrib , there is https://github.com/dotnet/coreclr/blob/master/Documentation/botr/porting-ryujit.md but it's a pretty high level overview. A community JIT port would certainly be very impressive.

**Comment by xied75** [comment]
@jasonwilliams200OK I did a search on source:

```
root@da1f6193b184:~/github/coreclr# find . -name "*.s"
```

Only managed to find two .s files?

```
./src/pal/tests/palsuite/composite/synchronization/nativecs_interlocked/sparcinterloc.s
./src/pal/tests/palsuite/composite/synchronization/nativecs_interlocked/hpitinterlock.s
```

Do we still somehow need to convert this first?

@benpye Something I can do as a small enough task?

**Comment by jasonwilliams200OK** [comment]
@xied75, try with capital `.S`. Building PAL on MIPS is the first step, which shouldn't require this conversion.

**Comment by benpye** [comment]
Porting PAL is reasonably straight forward, most of the architecture
specific files are in
https://github.com/dotnet/coreclr/tree/master/src/pal/src/arch though there
are various defines and ifdefs that may need updating.
https://github.com/dotnet/coreclr/blob/master/src/pal/src/exception/seh-unwind.cpp
has a lot of architecture specific code too. If you have more specific
questions I can try and give pointers from what I remember with porting PAL
to ARM.

**Comment by nidilong** [comment]
@ghost do you have some good news about MIPS support?

**Comment by geffzhang** [comment]
Any new developments?

**Comment by ievdokdm** [comment]
any updates on MIPS support?
I need it to run on MIPS 24Kc processor

**Comment by myFirstway** [comment]
any updates on MIPS support?

**Comment by zezba9000** [comment]
@myFirstway @ievdokdm  I'm working on a project (in early stages) to solve issues like this more quickly: https://github.com/reignstudios/IL2X

I can compile on MIPS right now but the runtime needs a lot more work done before its ready for general use.

**Comment by mms-** [comment]
Was about to use .Net Core for our embedded app, but this is a show stopper. On top of ARM32 and soon ARM64 we need MIPS support and that requirement doesn't seem to be changing anytime soon. Go and Rust do support MIPS.

**Comment by liujianm** [comment]
any updates on MIPS support?

**Comment by ibbcall** [comment]
any updates on MIPS support?

**Comment by BruceForstall** [comment]
Microsoft currently has no plans or work in progress to support MIPS. Of course, we would be willing to accept external contributions towards that goal as appropriate. Note that it is, certainly, a significant amount of work to port .NET Core to a new platform.

**Comment by theaoqi** [comment]
We (Loongson) are trying to port .NET Core (2.1 now, then will update to master) to MIPS64 and have just started this work recently. @liujianm @ibbcall

**Comment by am11** [comment]
@theaoqi, just FYI, the master branch builds with GCC as well as Clang, while <kbd>release/2.1</kbd> only builds with Clang. I am using master for SmartOS (Solaris) amd64 port work. :)

**Comment by redfox105** [comment]
> We (Loongson) are trying to port .NET Core (2.1 now, then will update to master) to MIPS64 and have just started this work recently. @liujianm @ibbcall

Is there a release plan?

**Comment by theaoqi** [comment]
> @theaoqi, just FYI, the master branch builds with GCC as well as Clang, while release/2.1 only builds with Clang. I am using master for SmartOS (Solaris) amd64 port work. :)

Thanks! I didn't know that and will consider this:)

**Comment by theaoqi** [comment]
> > We (Loongson) are trying to port .NET Core (2.1 now, then will update to master) to MIPS64 and have just started this work recently. @liujianm @ibbcall
> 
> Is there a release plan?

Not yet. Although we have some experience with porting OpenJDK to MIPS64, but we are not familiar with .NET. This is huge work, I cannot give a plan now.

**Comment by zhengwb** [comment]
> We (Loongson) are trying to port .NET Core (2.1 now, then will update to master) to MIPS64 and have just started this work recently. @liujianm @ibbcall

It is recommended to do this based on the latest version.

**Comment by liujianm** [comment]
> We (Loongson) are trying to port .NET Core (2.1 now, then will update to master) to MIPS64 and have just started this work recently. @liujianm @ibbcall

Thank you for the good news.
We want to run .netcore application on MIPS64(Loongson).
Hope to support in the latest version(3.0 will be released in September).
Looking forward to your good news.

**Comment by myFirstway** [comment]
> > > We (Loongson) are trying to port .NET Core (2.1 now, then will update to master) to MIPS64 and have just started this work recently. @liujianm @ibbcall
> > 
> > 
> > Is there a release plan?
> 
> Not yet. Although we have some experience with porting OpenJDK to MIPS64, but we are not familiar with .NET. This is huge work, I cannot give a plan now.

What's the progress now?

**Comment by kchanlee** [comment]
Looking forward to seeing .Net Core running on Loongson! @theaoqi Could you share the progress?

**Comment by theaoqi** [comment]
The code base was upgraded to 3.0. Hello World and serveral tests in coreclr can run on MIPS64 now. @kchanlee @myFirstway @liujianm

**Comment by BruceForstall** [comment]
@theaoqi It appears you are not pushing MIPS-related changes back to GitHub (or at least, not back to master)?

Are you using the interpreter, or are you porting RyuJIT?

Is there an overall MIPS porting plan documented here, in the coreclr repo?

Please be aware of the repository consolidation plan described here: https://github.com/dotnet/coreclr/issues/27549

**Comment by theaoqi** [comment]
> @theaoqi It appears you are not pushing MIPS-related changes back to GitHub (or at least, not back to master)?
> 

Not open souce yet :(

> Are you using the interpreter, or are you porting RyuJIT?

We skipped the interpreter and are porting JIT now. We found that interpreter doesn't work very well even on x86.

> 
> Is there an overall MIPS porting plan documented here, in the coreclr repo?

No.

> 
> Please be aware of the repository consolidation plan described here: #27549

Thank you very much for the information. I will be continuing concern over this issue. We plan to release 3.x (maybe 3.1) then catch up master. For 3.x, the repository consolidation doesn't affect us, right?

**Comment by BruceForstall** [comment]
> For 3.x, the repository consolidation doesn't affect us, right?

That's correct. The released product and servicing branches will remain as-is.

> We skipped the interpreter and are porting JIT now. We found that interpreter doesn't work very well even on x86.

Interesting. We've found in the past that using the interpreter allows us to bring up an architecture faster, or at least allows people to work in parallel, so everyone is not waiting on the JIT port to make progress. It's true that we haven't kept the interpreter building and being tested recently, so it has probably degraded. At one point it would successfully run well over 95% of our outerloop tests.

**Comment by theaoqi** [comment]
> > We skipped the interpreter and are porting JIT now. We found that interpreter doesn't work very well even on x86.
> 
> Interesting. We've found in the past that using the interpreter allows us to bring up an architecture faster, or at least allows people to work in parallel, so everyone is not waiting on the JIT port to make progress. It's true that we haven't kept the interpreter building and being tested recently, so it has probably degraded. At one point it would successfully run well over 95% of our outerloop tests.

Maybe we should do some research and test about interpreter again as our knowledge about coreclr and code quality are growing. Is pure interpreter supposed to work? Or dose it have to be interpreter and jit mixed? @xiangzhai may give more details.

**Comment by BruceForstall** [comment]
> Is pure interpreter supposed to work?

It should. However, it is not being actively tested right now, so I don't know how much "bit rot" needs to be repaired.

During bring-up, you do need to set a few variables, e.g.:
```
COMPlus_InterpreterDoLoopMethods=1   // allow the interpreter to run on functions with loops
COMPlus_GCgen0size=9999999 // prevent the GC from running very often; useful for bring-up
```

Check the other `COMPlus_Interpret*` variables in src\inc\clrconfigvalues.h -- some others might be necessary or useful.

**Comment by xiangzhai** [comment]
Hi @BruceForstall 

Please look at this issue  https://github.com/dotnet/coreclr/issues/24824

Thanks,
Leslie Zhai

**Comment by xiangzhai** [comment]
\cc @QiaoVanke

**Comment by whuanle** [comment]
What is the current progress?I will soon have a MIPS server where I need to run .NET Core programs.

**Comment by theaoqi** [comment]
> What is the current progress?I will soon have a MIPS server ,I need to run .NET Core programs.

please see https://github.com/dotnet/coreclr/issues/969#issuecomment-550129085
